python_modules/dagster-graphql/dagster_graphql/client/client.py

```diff
@@ -396,3 +400,189 @@ def terminate_run(self, run_id: str):
+                    }
+                else:
+                    metric_graphql_inputs[key]["jobMetricDefinitions"].append(
+                        {
+                            "metricValues": [
+                                {
+                                    "metricValue": metric_def.metric_value,
+                                    "metricName": metric_def.metric_name,
+                                }
+                                for metric_def in metric.metrics
+                            ],
+                        }
+                    )
+            else:
+                raise DagsterGraphQLClientError(
+                    "InvalidMetricTypeError",
+                    f"Metric {metric} is not of type AssetMetric or JobMetric",
+                )
+
+        res_data: Dict[str, Dict[str, Any]] = self._execute(
+            PUT_CLOUD_METRICS_MUTATION,
+            {"metrics": metric_graphql_inputs.values()},
+        )
+
+        store_metric_result: Dict[str, Any] = res_data["createExternalMetrics"]
+        store_metric_result_type: str = store_metric_result["__typename"]
+        if store_metric_result_type == "CreateExternalMetricsSuccess":
+            return
+
+        raise DagsterGraphQLClientError(store_metric_result_type, store_metric_result["message"])
+
+    @experimental
+    @public
+    def put_context_metrics(
+        self,
+        context: OpExecutionContext,
+        metrics: List[Metric],
+    ):
+        """Store metrics in the dagster cloud metrics store. This method is useful when you would like to
+        store run, asset or asset group materialization metric data to view in the insights UI.
+
+        Currently only supported in Dagster Cloud
+
+        Args:
+            metrics (Mapping[str, Any]): metrics to store in the dagster metrics store
+        """
+        check.list_param(metrics, "metrics", of_type=Metric)
+        check.inst_param(context, "context", OpExecutionContext)
+        metric_graphql_input = {}
+
+        if context.dagster_run.external_job_origin is None:
+            raise Exception("dagster run for this context has not started yet")
+
+        if context.has_assets_def:
+            metric_graphql_input = {
+                "assetMetricDefinitions": [
+                    {
+                        "runId": context.run_id,
+                        "stepKey": context.get_step_execution_context().step.key,
+                        "codeLocationName": context.dagster_run.external_job_origin.location_name,
+                        "repositoryName": (
+                            context.dagster_run.external_job_origin.external_repository_origin.repository_name
+                        ),
+                        "assetKey": selected_asset_key,
+                        "assetGroup": context.assets_def.group_names_by_key.get(
+                            selected_asset_key, None
+                        ),
+                        "metricValues": [
```

Comment:
Looks like this spreads the metrics evenly over all assets - we might want an alternate method that lets users allocate metrics per asset, with more specificity, having this approach as an alternate / simpler interface?

Reply:
can always ship in a follow-up, definitely non blocking